### ISSUE TITLE
RoundRobinDnsAddressResolverGroup causes memory leak in Pushy 0.13.5

### ISSUE THREAD

**Issue by eugeneseppel** [relayrides/pushy#715]
After several weeks running in production with #675 fix I've found a memory leak. 
There is a unit test that shows this leak:
`ApnsClientTest.java`, add this one:

```java
@Test
public void test1() throws Exception {
    final SimpleApnsPushNotification pushNotification
            = new SimpleApnsPushNotification("apns:1111111111111111111111111111111111111111111111111111111111111111",
            "com.twilio.www", "{}");

    NioEventLoopGroup group = new NioEventLoopGroup(2);
    for(int i = 0; ; i++) {
        final ApnsClient client = new ApnsClientBuilder()
                .setApnsServer("www.something.com") // to not flood apns
                .setSigningKey(signingKey)
                .setEventLoopGroup(group)
                .build();
        Thread.sleep(10);
        client.sendNotification(pushNotification).await();
        client.close().await();

        if (i % 50 == 0) {
            Runtime.getRuntime().gc();
        }

        System.out.printf("y=%d\n", i);
    }
}
```
Please run this test until 16000 iterations with current version of pushy and run

``` 
$ while true; do ps axu|grep test1|grep -v grep|awk '{print $5,$6}'; sleep 2; done
```

in terminal to check RSS is increasing
After that you can eliminate usage/creation of RoundRobinAddressResolver in ApnsChannelFactory:

>  remove private final AddressResolverGroup addressResolverGroup;
>          this.bootstrapTemplate.resolver(proxyHandlerFactory == null ?
>                  DefaultAddressResolverGroup.INSTANCE : NoopAddressResolverGroup.INSTANCE);

rerun this test and see that RSS is quite stable.

Moreover I have to add that currently in production our project consumes ~100 open UDP ports for resolver. (previously was ~1). It is not a serious issue for our usage pattern but quite strange why one need 100 ports to resolve just 2 APNS domain names 

Currently I use workaround that eliminates use (and creation) of RoundRobinDnsAddressResolverGroup in ApnsChannelFactory.java:
```
-        this.bootstrapTemplate.resolver(this.addressResolverGroup);
+        this.bootstrapTemplate.resolver(proxyHandlerFactory == null ?
+                DefaultAddressResolverGroup.INSTANCE : NoopAddressResolverGroup.INSTANCE);
```

**Comment by jchambers** [comment]
I can reproduce the memory issue and will investigate. I have not yet been able to reproduce the excessive open ports issue, though; can you provide more details?

Also, please accept my apologies for the delay in getting to this issue.

**Comment by jchambers** [comment]
I'm confident this is an upstream issue. I can reproduce it with a much more minimal test case:

```java
while (true) {
    final RoundRobinDnsAddressResolverGroup resolverGroup = new RoundRobinDnsAddressResolverGroup(
            NioDatagramChannel.class, DefaultDnsServerAddressStreamProvider.INSTANCE);

    resolverGroup.getResolver(CLIENT_EVENT_LOOP_GROUP.next());

    resolverGroup.close();
}
```

I'm working through some heap dumps to identify the leak now. I'll open an upstream issue (and, ideally, a pull request with a fix) as soon as I've pinned down the exact cause.

**Comment by jchambers** [comment]
I've filed the upstream issue as https://github.com/netty/netty/issues/10011 and proposed a fix at https://github.com/netty/netty/issues/10012. I don't think we have a great workaround within Pushy, and will need to wait for the issue to get resolved upstream before closing this issue.

**Comment by jchambers** [comment]
The upstream fix has been merged and will go out as part of Netty 4.1.46.

**Comment by eugeneseppel** [comment]
Hi Jon! Thank you for your investigation and bugfixes and sorry for my delay!
I'll check it too
